packages/matter.js/InterfaceNotes.ts

```diff
@@ -0,0 +1,121 @@
+// We agreed that server interface is the priority...  Here's a more detailed
+// proposal on the API.
+//
+// Note 1: This is a conceptual interface that models what the current PR is
+// generating (or will generate) for a single cluster.  I pulled it together
+// into a single place for discussion purposes.
+//
+// Question 1. We just call this "Device" but it's really a "ServerDevice".
+// In a controller we could potentially also have a "ClientDevice" that looks
+// very similar but proxies over the fabric to a remote endpoint.  So should
+// we consider making space in the API for a future Client/Server bifurcation
+// or am I thinking about this wrong?
+//
+// Note 2. The concept of a "ClientDevice" wouldn't be just for controllers, it
+// would also be for client clusters of a "ServerDevice".  I think.
+
+// ATTRIBUTES VALUES
+//
+// Attribute values are stored in an immutable "State" object.  This is not
+// necessarily the "source of truth".  It's just what the implementation
+// has conveyed to matter.js as the current attribute values.
+//
+// Note 3: (no longer relevant)
+//
+// Question 2: (resolved myself)
+type State = {
+    readonly currentLevel: number | undefined;
+    readonly minLevel?: number;
+    readonly maxLevel?: number;
+    readonly onOffTransitionTime?: number;
+    readonly onLevel: number | undefined;
+    readonly onTransitionTime?: number | undefined;
+    readonly offTransitionTime?: number | undefined;
+    readonly defaultMoveRate?: number | undefined;
+    readonly options: number;
+    readonly remainingTime?: number;
+    readonly startUpCurrentLevel?: number | undefined;
+}
+
+export interface PumpDevice extends Device {
+    // ATTRIBUTE MUTATION
+    //
+    // This is how the device implementation tells matter.js that attribute
+    // values are different.  Internal implementation would then read all
+    // attribute values and push any changes into the appropriate
+    // ClusterServer.
+    //
+    // Question 3: One reason I like this approach is I worry that notifying
+    // matter.js for every change would get too chatty.  Right now changes
+    // are coalesced in SubscriptionHandler before notifying subscribers so
+    // I don't think this is an issue from a network perspective.  But there
+    // are a lot of internal function calls for every attribute on every change
+    // I don't think there's any immediate need for change but is it worth
+    // considering for the shape of the public API for when there are hundreds
+    // of endpoints and thousands of attributes on a low-power device?
```

Comment:
Yeah, seems like there are a bunch of classes of attributes.  To test my understanding, here's the buckets I see:

# Session (volatile, writable by operator client)
# Configuration (non-volatile, writable by manager or administrator client)
# Reporting (writable by device)
# Fixed (writable at startup)

Fortunately all pretty well formalized in the spec.

Reply:
In fact to translate that to matter.js because we use other terminology and in some cases do not care if config or operational
* Attributes - can be get/set locally, remote only get, may be persisted
* WriteableAttribute - get/set locally AND from remote, may be persisted
* FixedAttributes - defined in default, local and remote get-only, never persisted
* FabricScopedWritableAttribute - get/set locally when knowing fabric AND from remote (fabric via session), always persisted

And yes somewhen we also add ACL checking and then we limit who can write what